docs(action-buttons): add guidelines regarding action button design conventions (#1290)
Lundalogik/lime-elements: src/design-guidelines/action-buttons/action-buttons.md

# Action buttons

> Action buttons are typically used in dialogs' footers, or in headers. These are buttons indicate to users that they must deliberately choose one action from a set of actions which are presented, for the system to continue or finish a task.

Author: I don't think _dialog's footer_ (or "dialogs' footers" in this case is grammatically wrong 🙃. Maybe "dialogs' footers" sounds a bit silly though. 
![image](https://user-images.githubusercontent.com/35954987/126750629-90f57933-1781-4859-a29c-edad1f7a3dc7.png)
[form here](https://developer.apple.com/design/human-interface-guidelines/macos/buttons/bevel-buttons/).

> Action buttons are typically used in dialogs' footers, or in headers. These are buttons indicate to users that they must deliberately choose one action from a set of actions which are presented, for the system to continue or finish a task.

Reviewer: No, it's not that it's grammatically wrong. It's just that "used in dialog footers" works just as well and is less clunky to say. The difference is easier to explain in Swedish:

- used in dialogs' footers -> används i dialogers sidfötter
- used in a dialog's footer -> används i en dialogs sidfot
- used in dialog footers -> används i dialogsidfötter

This is just to explain the relation between the different forms. In Swedish, I think the last form is the clunkiest, but in English, I think it's very neat.

"Used in a window's main content area" can't really be rewritten in the same way, because while 
(at least I think) you can call the footer of a dialog a "dialog footer", it doesn't really work to call the main content area of a window the "window main content area" (think of it as a single concatenated word in Swedish, like "windowmaincontentarea", vs "dialogfooter").

See the use of "dialog body" and "dialog footer" here:
![image](https://user-images.githubusercontent.com/154725/126753940-e61b3194-6b09-40f2-abf9-2401ae71557f.png)

> Action buttons are typically used in dialogs' footers, or in headers. These are buttons indicate to users that they must deliberately choose one action from a set of actions which are presented, for the system to continue or finish a task.

Reviewer: (You're perfectly free to choose which form to use of course, it's just a suggestion 😊)

Action buttons must naturally be presented together in one place; within the UI. When several action buttons are presented, their actions are often contradictory; for example **Discard** versus **Save**.

> When it comes to details such as placement of action buttons, choice of labels, and adding meaningful graphical details, it's important to follow a few design conventions, explained here:

Author: That's right. We can say "explained in this page" or "below"

> When it comes to details such as placement of action buttons, choice of labels, and adding meaningful graphical details, it's important to follow a few design conventions, explained here:

Reviewer: ```suggestion
When it comes to details such as placement of action buttons, choice of labels, and adding meaningful graphical details, it's important to follow a few design conventions, explained in this guide.
```

<br/>
<br/>

> *****

Author: Well, that's of course better. I thought we're usign `<hr>`s between sections of documentations as well, for instance between _Examples_ and _Properties_ sections. But they are just borders now when I check. So it should be fine to update Kompendiums styles for `<hr>`s

> *****

Reviewer: I was thinking more like add a border and some top margin to `<h2>`, and remove the need to manually add a `<hr>` every time.

Anyway, it's a nitpick and we don't have to do anything about it right now. It's better to get this documentation out there before we go on vacation.

> *****

Author: yes, good point. we can come back to it. ;)

# Action buttons

Action buttons are typically used in dialogs' footers, or in headers. These are buttons indicate to users that they must deliberately choose one action from a set of actions which are presented, for the system to continue or finish a task.

Action buttons must naturally be presented together in one place; within the UI. When several action buttons are presented, their actions are often contradictory; for example **Discard** versus **Save**.

When it comes to details such as placement of action buttons, choice of labels, and adding meaningful graphical details, it's important to follow a few design conventions, explained here:

<br/>
<br/>

*****

## Placement

> On a user interface which is designed for a left-to-right script like English, action buttons of a footer (for example in a dialog) are typically located at the bottom-right corner.

Author: a language cannpt be right-to-left. The way it is written can be :) I think that's why they chose to say script:

https://en.wikipedia.org/wiki/Right-to-left_script

But I don't know.

> On a user interface which is designed for a left-to-right script like English, action buttons of a footer (for example in a dialog) are typically located at the bottom-right corner.

Reviewer: Ok 👍

> On a user interface which is designed for a left-to-right script like English, action buttons of a footer (for example in a dialog) are typically located at the bottom-right corner.

Reviewer: In case you liked this suggestion. I think "of" is ok to though.

```suggestion
On a user interface which is designed for a left-to-right script like English, action buttons in a footer (for example in a dialog) are typically located at the bottom-right corner.
```

When there are two buttons placed together, the button that represents a positive action is always on the right side, and the one which presents a negative one is on the left.

A positive action could be an action that:
- takes the user forwards or to next step; e.g. *Next*, *Skip*, or *Continue*
- is primary, and most probably the one that users is going to select, e.g. *Save*, *Update*, *Send* (or even *Delete* which is a negative notion, but can be an expected or "primary" action, depending on the context, like in a confirm deletion dialog)

A negative action could be an action that:
- takes the user back, dismisses a dialog, or cancels an ongoing process; e.g. *Back*, *Cancel*, *Discard*
- is secondary, which means it is not likely that users clicks it in that context

<limel-example-action-buttons-placement></limel-example-action-buttons-placement>

:::tip

> In [`limel-dialog`](#/components/limel-dialog), there is a `slot` called `button`, which is flex-box. It automatically aligns its child elements to the right side. All you need to do is to write `slot="button"` on your elements.

Reviewer: As in this example, some of these suggestions contain two unrelated changes, that just happen to appear on the same line of code. You're obviously free to accept or reject them independently  (although you'll have to add the change manually if you wish to just add one and not the other).

<limel-example-action-buttons-primary-secondary></limel-example-action-buttons-primary-secondary>

A primary button has a stronger visual weight, and appearing more distinct makes it easier for users to chose the correct, safe, or promoted action.

You can use this visual distinction to promote the action that is safer, even if that action may not be the primary action. You can do so because sometimes:
- it can be hard to identify the intention of the user
- the action may be irreversible or risky

<limel-example-action-buttons-primary-secondary-reversed></limel-example-action-buttons-primary-secondary-reversed>

<br/>

*****

## Choosing good labels

Labels you use should make sense together. You are usually dealing with contradictory actions. Therefore, what you put on each button should also reflect the contradiction.

<limel-example-action-buttons-choosing-labels></limel-example-action-buttons-choosing-labels>

Labels must be making each option as clear as possible. An explicit label give users more confidence in selecting the correct action.

:::note

> - Make sure the label explains what the button does, rather than a generic label (such as “OK”, or "Yes").

Reviewer: This sentence uses different grammatical forms before and after the comma.

```suggestion
- Use a label that explains what the button does, rather than a generic label (such as “OK”, or "Yes").
```

- Using verbs as labels instead of nouns is better, because a verb describes the action better.
:::

<limel-example-action-buttons-choosing-explicit-labels></limel-example-action-buttons-choosing-explicit-labels>

<br/>

*****

## Providing a third alternative action

> Sometimes having two action is not enough for the user to take a decision with peace of mind. A common scenario is when a user has unsaved changes and tries to navigate away from the view. In such cases, typically a prompt maybe be shown to alert the user and give them a few logical choices.

Reviewer: ```suggestion
Sometimes having two actions is not enough for the user to make a decision with peace of mind. A common scenario is when a user has unsaved changes and tries to navigate away from the view. In such cases, typically a prompt maybe be shown to alert the user and give them a few logical choices.
```

> This prompt can offer two main actions of "Save" and "Discard" (which means "save & continue navigating away from this page" and "discard & continue navigating away from this page"), along with a third choice that allows users to "Cancel" (and go back / not continue navigating away!).

Reviewer: ```suggestion
This prompt can offer two main actions of "Save" and "Discard" (which means "save & continue navigating away from this page" and "discard & continue navigating away from this page"), along with a third choice that allows users to cancel (and go back / not continue navigating away!).
```

The quote marks and uppercase 'C' suggests "Cancel" as the actual label, like "Save" and "Discard", but "Cancel" isn't a good label for all the reasons outlined in the previous section 😄

Author: right 😄

<limel-example-action-buttons-third-alternative></limel-example-action-buttons-third-alternative>

> The action that takes users one step back inte the process is intentionally placed on the left side. This firstly makes it easier for users to mentally separate the other two, and secondly the placement makes it easier to connect the action to the notion of going back, in a UI which is designed for left-to-right script.

Reviewer: ```suggestion
The action that takes users one step back in the process is intentionally placed on the left side. This makes it easier for users to mentally separate the other two. Additionally, the placement makes it easier to connect the action to the notion of going back. (For left-to-right languages. Left and right will of course have reversed meaning in right-to-left languages, which is precisely why the CSS properties for these things use values like `flex-start` and `flex-end`, rather than `left` and `right`. See below.)
```

Author: scripts, not languages?

Reviewer: ```suggestion
The action that takes users one step back in the process is intentionally placed on the left side. This makes it easier for users to mentally separate the other two. Additionally, the placement makes it easier to connect the action to the notion of going back. (For left-to-right scripts. Left and right will of course have reversed meaning in right-to-left script, which is precisely why the CSS properties for these things use values like `flex-start` and `flex-end`, rather than `left` and `right`. See below.)
```

:::tip
If buttons are placed in a full width flex container that has `justify-content: flex-end;`, these styles will align the targeted button to left side of the flex box. See more [in examples of `limel-dialog`](#/components/limel-dialog).

```scss
.back-button {
    justify-self: flex-start;
    margin-right: auto;
}
```
:::

<br/>

*****

## Adding more meaning with colors and icons

> To make it easer for users to understand the message and take a decision faster, you can use icons and colors on the buttons.

Reviewer: ```suggestion
To make it easer for users to understand the message and make a decision faster, you can use icons and colors on the buttons.
```

Learn more about [usage of colors and color conventions in the UI](#/DesignGuidelines/color-system.md/), and [how to color limel-buttons](#/component/limel-button/).

<limel-example-action-buttons-icon-color></limel-example-action-buttons-icon-color>

:::note

> It might be tempting to use colors on action buttons, with the intention of making the user interface more cheerful. But using colors in a user interface must follow certain UI design conventions, as they convey meanings. Learn more about [usage of colors and color conventions in the UI](#/DesignGuidelines/color-system.md/).

Reviewer: ```suggestion
It might be tempting to use colors on action buttons to make the user interface more cheerful. But using colors in a user interface must follow certain UI design conventions, as colors convey meaning. Learn more about [usage of colors and color conventions in the UI](#/DesignGuidelines/color-system.md/).
```

:::

> <limel-example-action-buttons-primary-secondary-reversed-colors></limel-example-action-buttons-primary-secondary-reversed-colors>

Reviewer: It's not obvious how this examples relates to the surrounding text.

Author: Perhaps I should have included a "Don't" example for example with an unconventional color (instead of red) on a negative action button.

Reviewer: Ah, so, is this related to the note about not adding colours just to be cheerful?

Author: Yes. I was just trying to say that the orange button or the red button have their colors for a reason. E.g. Warning, Danger, etc... so don't just pick a random color to make things look nice. :)

Reviewer: Yeah, I think adding a "Don't" would help then. Also, adding a caption to each example would really help in understanding them. Like they do on material.io.

Reviewer: Or, frankly, I'd maybe just remove the examples. Even with a "Don't" example, I think the note is perfectly clear, and the examples don't actually help in understanding the message in this case. It's up to you though.

Author: Let's leave them there for now. I'll come up with a better clarification in future and update this guide.
